stdlib/Test/src/Test.jl

```diff
@@ -368,12 +370,73 @@ Test Passed
 This is equivalent to the uglier test `@test ≈(π, 3.14, atol=0.01)`.
 It is an error to supply more than one expression unless the first
 is a call expression and the rest are assignments (`k=v`).
+
+You can use any key for the `key=val` arguments, except for `broken` and `skip`,
+which have special meanings in the context of `@test`:
+
+* `broken=cond` indicates a test that should pass but currently consistently
+  fails when `cond==true`.  Tests that the expression `ex` evaluates to `false`
+  or causes an exception.  Returns a `Broken` `Result` if it does, or an `Error`
+  `Result` if the expression evaluates to `true`.  Regular `@test ex` is
+  evaluated when `cond==false`.
+* `skip=cond` marks a test that should not be executed but should be included in
+  test summary reporting as `Broken`, when `cond==true`.  This can be useful for
+  tests that intermittently fail, or tests of not-yet-implemented functionality.
+  Regular `@test ex` is evaluated when `cond==false`.
+
+# Examples
+
+```jldoctest
+julia> @test 2 + 2 ≈ 6 atol=1 broken=true
+Test Broken
+  Expression: ≈(2 + 2, 6, atol = 1)
+
+julia> @test 2 + 2 ≈ 5 atol=1 broken=false
+Test Passed
+
+julia> @test 2 + 2 == 5 skip=true
+Test Broken
+  Skipped: 2 + 2 == 5
+
+julia> @test 2 + 2 == 4 skip=false
+Test Passed
+```
+
+!!! compat "Julia 1.7"
+     The `broken` and `skip` keyword arguments require at least Julia 1.7.
 """
 macro test(ex, kws...)
+    # Collect the broken/skip keywords and remove them from the rest of keywords
+    broken = [kw.args[2] for kw in kws if kw.args[1] === :broken]
+    skip = [kw.args[2] for kw in kws if kw.args[1] === :skip]
+    kws = filter(kw -> kw.args[1] ∉ (:skip, :broken), kws)
+    # Validation of broken/skip keywords
+    for (kw, name) in ((broken, :broken), (skip, :skip))
+        if length(kw) > 1
+            error("invalid test macro call: cannot set $(name) keyword multiple times")
+        end
+    end
+    if length(skip) > 0 && length(broken) > 0
+        error("invalid test macro call: cannot set both skip and broken keywords")
+    end
+
+    # Build the test expression
     test_expr!("@test", ex, kws...)
     orig_ex = Expr(:inert, ex)
+
     result = get_test_result(ex, __source__)
-    :(do_test($result, $orig_ex))
+    testres = :(Broken(:skipped, $orig_ex))
+
+    return quote
+        if $(length(broken) > 0 && esc(broken[1]))
+            # code to call do_test with execution result and original expr
+            do_broken_test($result, $orig_ex)
+        elseif $(length(skip) > 0 && esc(skip[1]))
+            record(get_testset(), $testres)
+        else
+            do_test($result, $orig_ex)
+        end
+    end
```

Comment:
```suggestion
    testres = :(do_test($result, $orig_ex))
    if length(skip) > 0
        testres = :($(esc(skip[1])) ? Broken(:skipped, $orig_ex) : $testres)
    end
    if length(broken) > 0
        testres = :($(esc(broken[1])) ? do_broken_test($result, $orig_ex) : $testres)
    end
    return testres
```

Reply:
I think we want to avoid calling `do_test()` at all if `length(skip) > 0`.

Reply:
How about
```suggestion
    return quote
        if $(length(skip) > 0 && esc(skip[1]))
            record(get_testset(), Broken(:skipped, $orig_ex))
        else
            _do = $(length(broken) > 0 && esc(broken[1])) ? do_broken_test : do_test
            _do($result, $orig_ex)
        end
    end
```
which has a single copy of `result` as suggested by Jeff?